casanovo/data/db_utils.py

```diff
@@ -0,0 +1,317 @@
+            The maximum number of modifications to allow per peptide.
+        min_peptide_length : int
+            The minimum length of peptides to consider.
+        max_peptide_length : int
+            The maximum length of peptides to consider.
+
+        Returns
+        -------
+        mod_peptide_list : pd.DataFrame
+            A Pandas DataFrame with peptide, mass,
+            and protein columns. Sorted by neutral mass in ascending order.
+        """
+        # Verify the existence of the file:
+        if not os.path.isfile(fasta_filename):
+            logger.error("File %s does not exist.", fasta_filename)
+            raise FileNotFoundError(f"File {fasta_filename} does not exist.")
+
+        fasta_data = fasta.read(fasta_filename)
+        peptide_list = []
+        if digestion not in ["full", "partial"]:
+            logger.error("Digestion type %s not recognized.", digestion)
+            raise ValueError(f"Digestion type {digestion} not recognized.")
+        semi = digestion == "partial"
+        for header, seq in fasta_data:
+            pep_set = parser.cleave(
+                seq,
+                rule=parser.expasy_rules[enzyme],
+                missed_cleavages=missed_cleavages,
+                semi=semi,
+            )
+            protein = header.split()[0]
+            for pep in pep_set:
+                if (
+                    len(pep) < min_peptide_length
+                    or len(pep) > max_peptide_length
+                ):
+                    continue
+                if any(
+                    aa in pep for aa in "BJOUXZ"
+                ):  # Check for incorrect AA letters
+                    logger.warn(
+                        "Skipping peptide with ambiguous amino acids: %s", pep
+                    )
+                    continue
+                peptide_list.append((pep, protein))
+
+        # Generate modified peptides
+        mass_calculator = depthcharge.masses.PeptideMass(residues="massivekb")
+        mod_peptide_list = []
+        for pep, prot in peptide_list:
+            peptide_isoforms = parser.isoforms(
+                pep,
+                variable_mods=self.var_mods,
+                fixed_mods=self.fixed_mods,
+                max_mods=max_mods,
+            )
+            peptide_isoforms = list(
+                map(ProteinDatabase._convert_from_modx, peptide_isoforms)
+            )
+            mod_peptide_list.extend(
+                [mod_pep, mass_calculator.mass(mod_pep), prot]
+                for mod_pep in peptide_isoforms
+            )
+
+        # Create a DataFrame for easy sorting and filtering
+        pdb_df = pd.DataFrame(
+            mod_peptide_list, columns=["peptide", "calc_mass", "protein"]
+        )
+        pdb_df.sort_values(by=["calc_mass", "peptide"], inplace=True)
+
+        logger.info("Digestion complete. %d peptides generated.", len(pdb_df))
+        return pdb_df
+
+    def _construct_mods_dict(self, allowed_mods):
```

Comment:
**todo:** I don't really like how this is handled now. It's not consistent with the AA alphabet, it doesn't use a flexible notation nor the same notation as modifications elsewhere.

Reply:
https://pyteomics.readthedocs.io/en/latest/api/parser.html#pyteomics.parser.isoforms

Using pyteomics requires adding these lowercase prefixes in a dictionary so that it can generate isoforms. I looked online but could not find consistent namings for modifications (https://pyteomics.readthedocs.io/en/latest/parser.html). Either way, it is simply an intermediate step that allows me to use pyteomics to generate isoforms, then I immediately convert from modX format back to the mod format Casanovo uses.

Do you have any suggestions as to how I can change this?

Reply:
So the main thing that annoys me is that this mixes configuration with implementation, with a bunch of hardcoded values that mirror the config file. If something changes there, this will get out of sync, resulting in pernicious bugs. For example, with the new version of DepthCharge, we're switching to ProForma for modifications rather than MassIVE-KB style, which will make this code completely non-functional.

So this needs to be more flexible, and ideally also a better translation between Casanovo modification notation and Pyteomics modification notation.

Maybe some functionality from the DepthCharge alphabet handling can be re-used for this (in the new version).

Reply:
Made the consideration of modifications in db search mode not hardcoded. Now uses/parses a custom modification input schema like TIDE. This is easily interchangeable with any further vocabulary changes.